game/games/snake/canvas.gd

```diff
@@ -0,0 +1,166 @@
+func _process(_delta):
+	_tick += 1
+	if _tick == 4:
+		_tick = 0
+		var new_head_pos = _move_snake()
+		var collision = _check_for_collision(new_head_pos)
+		if collision:
+			_game_over()
+		else:
+			_check_for_apple(new_head_pos)
+		update()  # triggers NOTIFICATION_DRAW of CanvasItem / _draw()
+
+
+func _draw():
+	_draw_snake()
+	_draw_apple()
+
+
+func _input(_event):
+	if Input.is_action_just_pressed("ui_up"):
+		_set_game_paused(false)
+		if _direction != Direction.DOWN:
+			_direction = Direction.UP
+	elif Input.is_action_just_pressed("ui_down"):
+		_set_game_paused(false)
+		if _direction != Direction.UP:
+			_direction = Direction.DOWN
+	elif Input.is_action_just_pressed("ui_left"):
+		_set_game_paused(false)
+		if _direction != Direction.RIGHT:
+			_direction = Direction.LEFT
+	elif Input.is_action_just_pressed("ui_right"):
+		_set_game_paused(false)
+		if _direction != Direction.LEFT:
+			_direction = Direction.RIGHT
+
+
+func _calc_tile_size():
+	var spacing = Vector2(white_space * 2, white_space * 2)
+	var slice = (OS.get_window_size() - spacing) / tile_count
+	var size = 0
+	if slice.x > slice.y:
+		size = slice.y
+	else:
+		size = slice.x
+	_tile_size = int(floor(size))
+
+
+func _set_playfield():
+	var size = _tile_size * tile_count
+	get_parent().rect_min_size = Vector2(size, size)
+
+
+func _set_game_paused(value: bool):
+	set_process(not value)
+	_game_paused = value
+
+
+func _set_snake():
+	_snake.clear()
+	var center = int(tile_count / 2)
+	_snake.push_back(Vector2(center, center))
+	_direction = null
+
+
+func _set_apple():
+	var x = randi() % tile_count
+	var y = randi() % tile_count
+	_apple_pos = Vector2(float(x), float(y))
+
+
+func _move_snake() -> Vector2:
+	var snake_head = _snake.back()
+	var new_head_pos = snake_head
+	if _direction == Direction.UP:
+		new_head_pos.y -= 1
+	elif _direction == Direction.DOWN:
+		new_head_pos.y += 1
+	elif _direction == Direction.LEFT:
+		new_head_pos.x -= 1
+	elif _direction == Direction.RIGHT:
+		new_head_pos.x += 1
+	return new_head_pos
+
+
+func _check_for_collision(new_head_pos: Vector2) -> bool:
+	for snake_part in _snake:
+		if new_head_pos.is_equal_approx(snake_part):
+			return true
+	if new_head_pos.x < 0:
+		return true
+	if new_head_pos.x >= tile_count:
+		return true
+	if new_head_pos.y < 0:
+		return true
+	if new_head_pos.y >= tile_count:
+		return true
+	return false
+
+
+func _game_over():
+	_set_game_paused(true)
+	_set_snake()
+	if _score > _high_score:
+		_high_score = _score
+	_score = 0
+	_main.set_score(_score)
+	_main.set_highscore(_high_score)
+	GameManager.end_game("Ouch! You got a score of %s!" % _high_score, _high_score)
+
+
+func _check_for_apple(new_head_pos: Vector2):
+	_snake.push_back(new_head_pos)
+	if new_head_pos.is_equal_approx(_apple_pos):
+		_set_apple()
+		_score += 1
+		_main.set_score(_score)
+	else:
+		_snake.remove(0)
+
+
+func _draw_snake():
+	var color = Color(0.016, 1, 0)
+	for snake_part in _snake:
+		_draw_square(snake_part, color)
```

Comment:
I think the head should have a different colour

Reply:
`_draw_square(_snake.back(), Color(0, 0.5, 0))`
makes it look better

Reply:
then what about a gradient on the snake?
e.g. something like the following:
```gd
func _draw_snake():
	var i := 0.0
	for snake_part in _snake:
		i += 0.5 / _snake.size()
		_draw_square(snake_part, Color(0.016, 0.5 + i, 0))
```

For me personally the snake currently looks boring